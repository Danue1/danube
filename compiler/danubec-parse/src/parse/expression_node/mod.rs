mod literal_expression_node;

use crate::State;

impl crate::Context {
    pub fn expression_node(&mut self) -> State {
        one_of!(self.expression_literal_node())
    }
}
