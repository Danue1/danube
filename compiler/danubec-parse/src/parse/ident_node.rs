use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn ident_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, IDENT_NODE);
        self.finish_node();
        State::Stop
    }
}
