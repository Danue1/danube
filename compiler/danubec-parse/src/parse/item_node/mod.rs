mod enum_item_node;
mod struct_item_node;

impl crate::Context {
    pub fn item_node(&mut self) -> crate::State {
        one_of!(self.struct_item_node(), self.enum_item_node())
    }
}
