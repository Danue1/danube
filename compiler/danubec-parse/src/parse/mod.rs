mod ident_node;
mod item_node;
mod path_node;
mod type_node;

use crate::context::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::context::Context {
    #[inline]
    pub fn ast(&mut self) {
        self.start_node(SyntaxKind::AST_NODE);
        self.skip_whitespace();
        while !self.is_empty() {
            if self.item_node() == State::Continue {
                self.unexpected_token();
            }
            self.skip_whitespace();
        }
        self.finish_node();
    }

    fn unexpected_token(&mut self) {
        self.start_node(SyntaxKind::ERROR);
        self.bump();
        self.finish_node();
    }
}
