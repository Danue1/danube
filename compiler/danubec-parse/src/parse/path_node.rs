use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn path_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, PATH_NODE, ident_node);

        while !self.is_eof() {
            if expect!(self, SyntaxKind::COLON) {
                if expect!(self, SyntaxKind::COLON) {
                    self.skip_whitespace();
                    self.ident_node();
                } else {
                    self.unexpected_token();
                }
            } else {
                break;
            }
        }

        self.finish_node();
        State::Stop
    }
}
