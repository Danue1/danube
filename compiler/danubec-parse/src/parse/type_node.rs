use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn type_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.unnamed_type_kind_node() == State::Continue
            && self.path_type_kind_node() == State::Continue
        {
            State::Continue
        } else {
            self.start_node_at(checkpoint, SyntaxKind::TYPE_NODE);
            self.finish_node();
            State::Stop
        }
    }

    // ()
    // (type1)
    // (type1, type2)
    pub fn unnamed_type_kind_node(&mut self) -> State {
        guard!(self, LEFT_PAREN, UNNAMED_TYPE_KIND_NODE);
        self.skip_whitespace();

        while !self.is_eof() {
            if expect!(self, SyntaxKind::RIGHT_PAREN) {
                break;
            }

            let checkpoint = self.checkpoint();
            if self.type_node() == State::Continue {
                self.unexpected_token();
            } else {
                self.start_node_at(checkpoint, SyntaxKind::UNNAMED_TYPE_KIND_NODE);
                self.finish_node();
            }
            self.skip_whitespace();
            if expect!(self, SyntaxKind::COMMA) {
                self.skip_whitespace();
            }
        }

        self.finish_node();
        State::Stop
    }

    // Foo
    // Foo::Bar
    pub fn path_type_kind_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.path_node() == State::Continue {
            return State::Continue;
        }

        self.start_node_at(checkpoint, SyntaxKind::PATH_TYPE_KIND_NODE);
        self.finish_node();
        State::Stop
    }
}
