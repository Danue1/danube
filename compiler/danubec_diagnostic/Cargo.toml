[package]
name = "danubec_diagnostic"
version = "0.1.0"
edition = "2024"

[dependencies]
