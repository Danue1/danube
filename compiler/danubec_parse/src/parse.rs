use crate::{
    event::Event,
    grammar::{Context, krate},
};
use danubec_diagnostic::Diagnostic;
use danubec_lex::lex;
use danubec_syntax::{GreenNodeBuilder, SyntaxKind, SyntaxNode};

pub fn parse(source: &str, diagnostic: &mut Diagnostic) -> SyntaxNode {
    let mut parse = |tokens: &[_]| {
        let tokens: Vec<_> = tokens.iter().map(|&(kind, _)| kind).collect();
        let mut context = Context::new(&tokens, diagnostic);
        krate(&mut context);

        context.finish()
    };

    let tokens = lex(source);
    let events = parse(&tokens);

    build(&tokens, events)
}

fn build(mut tokens: &[(SyntaxKind, &str)], mut events: Vec<Event>) -> SyntaxNode {
    let mut builder = GreenNodeBuilder::new();

    macro_rules! advance {
        () => {
            if let Some(&(kind, text)) = tokens.get(0) {
                builder.token(kind.into(), text);
                tokens = &tokens[1..];
            }
        };
    }

    macro_rules! trivia {
        () => {
            loop {
                match tokens.get(0) {
                    Some((kind, _)) if kind.at_trivia() => advance!(),
                    _ => break,
                }
            }
        };
    }

    for index in 0..events.len() {
        match std::mem::replace(&mut events[index], Event::Placeholder) {
            event @ Event::Start { forward_parent, .. }
            | event @ Event::Expire { forward_parent } => {
                let mut forward_parents = match event {
                    Event::Start { kind, .. } => vec![kind],
                    _ => vec![],
                };
                let mut index = index;
                let mut forward_parent = forward_parent;

                while let Some(fp) = forward_parent {
                    index += fp;
                    match std::mem::replace(&mut events[index], Event::Placeholder) {
                        Event::Start {
                            kind,
                            forward_parent: fp,
                        } => {
                            forward_parents.push(kind);
                            forward_parent = fp;
                        }
                        Event::Expire { forward_parent: fp } => {
                            forward_parent = fp;
                        }
                        _ => {
                            //
                        }
                    }
                }

                for kind in forward_parents.into_iter().rev() {
                    builder.start_node(kind.into());
                }

                trivia!();
            }
            Event::Token => {
                advance!();
                trivia!();
            }
            Event::End => builder.finish_node(),
            Event::Placeholder => {
                //
            }
        }
    }

    SyntaxNode::new_root(builder.finish())
}
