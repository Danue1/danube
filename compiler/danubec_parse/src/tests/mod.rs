use crate::parse;
use danubec_diagnostic::Diagnostic;

#[test]
fn top_level_attribute() {
    let source = r#"
#![a("value")]
#![a(key)]
#![a(key1, key2)]
#![a(key = "value")]
#![a::b(key = "value")]"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn attribute() {
    let source = r#"
#[a("value")]
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn zero_sized_struct() {
    let source = r#"
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn named_struct() {
    let source = r#"
struct Foo { }
struct Bar {
    a: i32,
    b: String,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn unnamed_struct() {
    let source = r#"
struct Foo();
struct Bar(());
struct Baz(i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn mixed_enum() {
    let source = r#"
enum Foo {}
enum Bar {
    A,
    B { a: i32, b: String },
    C(i32, String),
    D = 42,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn module() {
    let source = r#"
mod foo;
mod bar {
    struct Foo;
    enum Bar {
        A,
        B { a: i32, b: String },
        C(i32, String),
        D = 42,
    }
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn constant() {
    let source = r#"
const FOO: i32 = 42;
const BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn static_variable() {
    let source = r#"
static FOO: i32 = 42;
static BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn type_alias() {
    let source = r#"
type Foo = i32;
type Bar = (i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn function() {
    let source = r#"
fn foo();
fn bar() {}
fn baz(a: i32, b: String) -> (i32, String) {
    (a, b)
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn trait_definition() {
    let source = r#"
trait Foo { }
trait Bar {
    fn foo();
    const BAR: i32;
    type Baz;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn anonymous_implement() {
    let source = r#"
impl Foo { }
impl Foo {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn targeted_implement() {
    let source = r#"
impl Foo for Bar { }
impl Foo for Bar {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn use_definition() {
    let source = r#"
use {};
use {{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use *;
use foo;
use foo as bar;
use foo::{};
use foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};

use ::{};
use ::{{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use ::*;
use ::foo;
use ::foo as bar;
use ::foo::{};
use ::foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};
"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}
