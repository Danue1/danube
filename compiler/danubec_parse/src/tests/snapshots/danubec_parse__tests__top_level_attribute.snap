---
source: compiler/danubec_parse/src/tests/mod.rs
expression: node
---
KRATE_NODE@0..90
  NEW_LINE@0..1 "\n"
  TOP_LEVEL_ATTRIBUTE_NODE@1..16
    HASH@1..2 "#"
    EXCLAMATION@2..3 "!"
    LEFT_BRACKET@3..4 "["
    NESTED_ATTRIBUTE_ARGUMENT_NODE@4..14
      PATH_NODE@4..5
        PATH_SEGMENT_IDENTIFIER_NODE@4..5
          IDENTIFIER_NODE@4..5
            IDENTIFIER_SEGMENT@4..5
              IDENTIFIER@4..5 "a"
      LEFT_PAREN@5..6 "("
      ATTRIBUTE_ARGUMENT_NODE@6..13
        LITERAL_EXPRESSION_NODE@6..13
          STRING_LITERAL_NODE@6..13
            STRING_START@6..7 "\""
            STRING_LITERAL_TEXT_NODE@7..12
              STRING_SEGMENT@7..12 "value"
            STRING_END@12..13 "\""
      RIGHT_PAREN@13..14 ")"
    RIGHT_BRACKET@14..15 "]"
    NEW_LINE@15..16 "\n"
  TOP_LEVEL_ATTRIBUTE_NODE@16..27
    HASH@16..17 "#"
    EXCLAMATION@17..18 "!"
    LEFT_BRACKET@18..19 "["
    NESTED_ATTRIBUTE_ARGUMENT_NODE@19..25
      PATH_NODE@19..20
        PATH_SEGMENT_IDENTIFIER_NODE@19..20
          IDENTIFIER_NODE@19..20
            IDENTIFIER_SEGMENT@19..20
              IDENTIFIER@19..20 "a"
      LEFT_PAREN@20..21 "("
      KEY_VALUE_ATTRIBUTE_ARGUMENT_NODE@21..24
        PATH_NODE@21..24
          PATH_SEGMENT_IDENTIFIER_NODE@21..24
            IDENTIFIER_NODE@21..24
              IDENTIFIER_SEGMENT@21..24
                IDENTIFIER@21..24 "key"
      RIGHT_PAREN@24..25 ")"
    RIGHT_BRACKET@25..26 "]"
    NEW_LINE@26..27 "\n"
  TOP_LEVEL_ATTRIBUTE_NODE@27..45
    HASH@27..28 "#"
    EXCLAMATION@28..29 "!"
    LEFT_BRACKET@29..30 "["
    NESTED_ATTRIBUTE_ARGUMENT_NODE@30..43
      PATH_NODE@30..31
        PATH_SEGMENT_IDENTIFIER_NODE@30..31
          IDENTIFIER_NODE@30..31
            IDENTIFIER_SEGMENT@30..31
              IDENTIFIER@30..31 "a"
      LEFT_PAREN@31..32 "("
      KEY_VALUE_ATTRIBUTE_ARGUMENT_NODE@32..36
        PATH_NODE@32..36
          PATH_SEGMENT_IDENTIFIER_NODE@32..36
            IDENTIFIER_NODE@32..36
              IDENTIFIER_SEGMENT@32..36
                IDENTIFIER@32..36 "key1"
      COMMA@36..37 ","
      WHITESPACE@37..38 " "
      KEY_VALUE_ATTRIBUTE_ARGUMENT_NODE@38..42
        PATH_NODE@38..42
          PATH_SEGMENT_IDENTIFIER_NODE@38..42
            IDENTIFIER_NODE@38..42
              IDENTIFIER_SEGMENT@38..42
                IDENTIFIER@38..42 "key2"
      RIGHT_PAREN@42..43 ")"
    RIGHT_BRACKET@43..44 "]"
    NEW_LINE@44..45 "\n"
  TOP_LEVEL_ATTRIBUTE_NODE@45..66
    HASH@45..46 "#"
    EXCLAMATION@46..47 "!"
    LEFT_BRACKET@47..48 "["
    NESTED_ATTRIBUTE_ARGUMENT_NODE@48..64
      PATH_NODE@48..49
        PATH_SEGMENT_IDENTIFIER_NODE@48..49
          IDENTIFIER_NODE@48..49
            IDENTIFIER_SEGMENT@48..49
              IDENTIFIER@48..49 "a"
      LEFT_PAREN@49..50 "("
      KEY_VALUE_ATTRIBUTE_ARGUMENT_NODE@50..63
        PATH_NODE@50..54
          PATH_SEGMENT_IDENTIFIER_NODE@50..54
            IDENTIFIER_NODE@50..54
              IDENTIFIER_SEGMENT@50..54
                IDENTIFIER@50..53 "key"
                WHITESPACE@53..54 " "
        EQUAL@54..55 "="
        WHITESPACE@55..56 " "
        LITERAL_EXPRESSION_NODE@56..63
          STRING_LITERAL_NODE@56..63
            STRING_START@56..57 "\""
            STRING_LITERAL_TEXT_NODE@57..62
              STRING_SEGMENT@57..62 "value"
            STRING_END@62..63 "\""
      RIGHT_PAREN@63..64 ")"
    RIGHT_BRACKET@64..65 "]"
    NEW_LINE@65..66 "\n"
  TOP_LEVEL_ATTRIBUTE_NODE@66..90
    HASH@66..67 "#"
    EXCLAMATION@67..68 "!"
    LEFT_BRACKET@68..69 "["
    NESTED_ATTRIBUTE_ARGUMENT_NODE@69..88
      PATH_NODE@69..73
        PATH_SEGMENT_IDENTIFIER_NODE@69..70
          IDENTIFIER_NODE@69..70
            IDENTIFIER_SEGMENT@69..70
              IDENTIFIER@69..70 "a"
        COLON@70..71 ":"
        COLON@71..72 ":"
        PATH_SEGMENT_IDENTIFIER_NODE@72..73
          IDENTIFIER_NODE@72..73
            IDENTIFIER_SEGMENT@72..73
              IDENTIFIER@72..73 "b"
      LEFT_PAREN@73..74 "("
      KEY_VALUE_ATTRIBUTE_ARGUMENT_NODE@74..87
        PATH_NODE@74..78
          PATH_SEGMENT_IDENTIFIER_NODE@74..78
            IDENTIFIER_NODE@74..78
              IDENTIFIER_SEGMENT@74..78
                IDENTIFIER@74..77 "key"
                WHITESPACE@77..78 " "
        EQUAL@78..79 "="
        WHITESPACE@79..80 " "
        LITERAL_EXPRESSION_NODE@80..87
          STRING_LITERAL_NODE@80..87
            STRING_START@80..81 "\""
            STRING_LITERAL_TEXT_NODE@81..86
              STRING_SEGMENT@81..86 "value"
            STRING_END@86..87 "\""
      RIGHT_PAREN@87..88 ")"
    RIGHT_BRACKET@88..89 "]"
    NEW_LINE@89..90 "\n"
